jenkins/Jenkinsfile

```diff
@@ -1,374 +1,492 @@
+                            sh """
+                                sed -i "s/$image.*/$image:$version\\"/" jenkins/gce-custom-images/prepare_custom_leonardo_gce_image.sh
+                            """
+                        }
+
+                        versionMap[image] = version
+                    }
+
+                    if (GENERATE_DATAPROC_IMAGE == "true") {
+                        sh """
+                            printf 'Printing any changed versions in prepare-custom-leonardo-jupyter-dataproc-image.sh:\\n\\n'
+                            git diff jenkins/dataproc-custom-images/prepare-custom-leonardo-jupyter-dataproc-image.sh
+                        """
+                    }
+
+                    if (GENERATE_GCE_IMAGE == "true") {
+                        sh """
+                            printf 'Printing any changed versions in prepare_custom_leonardo_gce_image.sh:\\n\\n'
+                            git diff jenkins/gce-custom-images/prepare_custom_leonardo_gce_image.sh
+                        """
+                    }
+                }
+            }
     }
 
-      stage('Run build script') {
+    stage('Run image generation scripts') {
         when { expression { !shouldAbortBuild } }
-        steps {
-          sh """
-              git submodule init && git submodule update
-          """
-          script {
-            for(String dp_version: dataproc_versions.trim().split(' ')) {
-              def dp_version_formatted = dp_version.replace(".","-")
-              // Dataproc 1.2 is used for the legacy image and must have python 3.6 installed
-              // Otherwise, a terra-docker image is used which must have python 3.7 installed
-              def python_version = dp_version.startsWith("1.2") ? "3.6.8" : "3.7.4"
-              println("Installing Python $python_version on Dataproc $dp_version")
-              sh (
-                script: """
-                  sed -i 's/python_version=.*/python_version="$python_version"/' jenkins/dataproc-custom-images/prepare-custom-leonardo-jupyter-dataproc-image.sh
-                  cat jenkins/dataproc-custom-images/prepare-custom-leonardo-jupyter-dataproc-image.sh
-                  cd jenkins/dataproc-custom-images/dataproc-custom-images
-                  python generate_custom_image.py \
-                    --image-name "$IMAGE_BASE_NAME-$dp_version_formatted-$imageID" \
-                    --dataproc-version "$dp_version" \
-                    --customization-script ../prepare-custom-leonardo-jupyter-dataproc-image.sh \
-                    --zone us-central1-a \
-                    --gcs-bucket $BUCKET_NAME \
-                    --project-id=$GOOGLE_PROJECT \
-                    --disk-size=50
-                """
-              )
+            // Fail the pipeline if any of the parallel branches (that build the Dataproc and GCE images) fails
+            failFast true
+            parallel {
```

Comment:
how do the logs look when this is run in parallel? Is it easy to identify 'which is which'

Reply:
Interleaved like this

```
[Run Dataproc image generation script] Uploading files to GCS bucket.
[Run Dataproc image generation script] + gsutil cp ../prepare-custom-leonardo-jupyter-dataproc-image.sh gs://leo-dataproc-image/custom-image-custom-leo-image-dataproc-1-2-79-debian9-54076f2-20200304-135245/sources/init_actions.sh
[Run GCE image generation script] [custom-leonardo-gce-image-generator]: 2020-03-04T18:52:46Z Validation Complete
[Run GCE image generation script] [custom-leonardo-gce-image-generator]: 2020-03-04T18:52:46Z Workflow Project: broad-dsp-gcr-public
[Run GCE image generation script] [custom-leonardo-gce-image-generator]: 2020-03-04T18:52:46Z Workflow Zone: us-central1-a
[Run GCE image generation script] [custom-leonardo-gce-image-generator]: 2020-03-04T18:52:46Z Workflow GCSPath: gs://leo-gce-image
[Run GCE image generation script] [custom-leonardo-gce-image-generator]: 2020-03-04T18:52:46Z Daisy scratch path: https://console.cloud.google.com/storage/browser/leo-gce-image/daisy-custom-leonardo-gce-image-generator-20200304-18:52:44-m39nv
[Run GCE image generation script] [custom-leonardo-gce-image-generator]: 2020-03-04T18:52:46Z Uploading sources
[Run GCE image generation script] [custom-leonardo-gce-image-generator]: 2020-03-04T18:52:46Z Running workflow
[Run GCE image generation script] [custom-leonardo-gce-image-generator]: 2020-03-04T18:52:46Z Running step "create-disks" (CreateDisks)
[Run GCE image generation script] [custom-leonardo-gce-image-generator.create-disks]: 2020-03-04T18:52:46Z CreateDisks: Creating disk "disk-install-custom-leonardo-gce-image-generator-m39nv".
[Run Dataproc image generation script] Copying file://../prepare-custom-leonardo-jupyter-dataproc-image.sh [Content-Type=application/x-sh]...
[Run Dataproc image generation script] / [0 files][    0.0 B/  7.1 KiB]                                                
/ [1 files][  7.1 KiB/  7.1 KiB]                                                
[Run Dataproc image generation script] Operation completed over 1 objects/7.1 KiB.
```
which is not great but when I needed to searching for logs from a specific one I text-searched  with `Run Dataproc image generation script` which highlighted the lines of interest in yellow and it wasn't too bad. Not running these in parallel would incur too much time penalty I think.